drivers/usb/udc/udc_dwc2_vendor_quirks.h

```diff
@@ -148,6 +149,180 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_STM32F4_FSOTG_DEFINE)
+	}
+}
+
+static inline int usbhs_enable_nrfs_service(const struct device *dev)
+{
+	nrfs_err_t nrfs_err;
+	int err;
+
+	err = nrfs_backend_wait_for_connection(K_MSEC(1000));
+	if (err) {
+		LOG_INF("NRFS backend connection timeout");
+		return err;
+	}
+
+	nrfs_err = nrfs_usb_init(usbhs_vbus_handler);
+	if (nrfs_err != NRFS_SUCCESS) {
+		LOG_ERR("Failed to init NRFS VBUS handler: %d", nrfs_err);
+		return -EIO;
+	}
+
+	nrfs_err = nrfs_usb_enable_request((void *)dev);
+	if (nrfs_err != NRFS_SUCCESS) {
+		LOG_ERR("Failed to enable NRFS VBUS service: %d", nrfs_err);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static inline int usbhs_enable_core(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	wrapper->ENABLE = USBHS_ENABLE_PHY_Msk | USBHS_ENABLE_CORE_Msk;
+	wrapper->TASKS_START = 1UL;
+
+	/* Enable interrupt */
+	wrapper->INTENSET = 1UL;
+
+	return 0;
+}
+
+static inline int usbhs_disable_core(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	wrapper->ENABLE = 0UL;
+	wrapper->TASKS_START = 1UL;
+
+	/* Enable interrupt */
+	wrapper->INTENSET = 0UL;
+
+	return 0;
+}
+
+static inline int usbhs_disable_nrfs_service(const struct device *dev)
+{
+	nrfs_err_t nrfs_err;
+
+	nrfs_err = nrfs_usb_disable_request((void *)dev);
+	if (nrfs_err != NRFS_SUCCESS) {
+		LOG_ERR("Failed to disable NRFS VBUS service: %d", nrfs_err);
+		return -EIO;
+	}
+
+	nrfs_usb_uninit();
+
+	return 0;
+}
+
+static inline int usbhs_irq_clear(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	wrapper->EVENTS_CORE = 0UL;
+
+	return 0;
+}
+
+static inline int usbhs_init_caps(const struct device *dev)
+{
+	const struct udc_dwc2_config *config = dev->config;
+	struct udc_data *data = dev->data;
+	int err;
+
+	data->caps.can_detect_vbus = true;
+	data->caps.hs = true;
+
+	/* Register nRF54H20 OUT endpoints 1 ... 5 and 12 ... 15 */
+	for (size_t i = 1; i < config->num_out_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		config->ep_cfg_out[i].caps.bulk = 1;
+		config->ep_cfg_out[i].caps.interrupt = 1;
+		config->ep_cfg_out[i].caps.iso = 1;
+		config->ep_cfg_out[i].caps.mps = 1024U;
+		if (i < 6) {
+			config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		} else {
+			config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | (i + 6);
+		}
+
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	/* Register nRF54H20 IN endpoints 1 ... 11 */
+	for (size_t i = 1; i < config->num_in_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		config->ep_cfg_in[i].caps.bulk = 1;
+		config->ep_cfg_in[i].caps.interrupt = 1;
+		config->ep_cfg_in[i].caps.iso = 1;
+		config->ep_cfg_in[i].caps.mps = 1024U;
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	return 0;
+}
```

Comment:
I believe it would be more appropriate for the quirk to simply return `GHWCFG1` value (here: `USBHSCORE_GHWCFG1_ResetValue`) and make the stack to configure the endpoint directions based on it.

Reply:
We can add required GHWCFGn properties to the bindings and then have it generic for all the vendor implementations.

Reply:
That would work as well.